File: Week2/SortingProblem/InsertionSort.java
public class InsertionSort {
/*
 * The task is to complete the insert() function which is used to implement Insertion Sort.
 * https://www.geeksforgeeks.org/problems/insertion-sort/1
 * T(c) -> O(n^2)  // bast case -> O(n)
 * S(c) -> O(1)
 * Stable sort - A sorting algorithm is considered stable if it preserves the relative order of equal elements in the sorted output as they were in the original input. 
 */
    static void insert(int arr[],int i)
    {
        int temp = arr[i];
        int j = i-1;
        
        while(j>=0 && arr[j]>temp)
        {
        // shift element 1 place right
            arr[j+1] = arr[j];
            j--;
        }
        
        arr[j+1]  = temp;
    }

    //Function to sort the array using insertion sort algorithm.
    public static void insertionSort(int arr[], int n)
    {
        //code here
        for(int i=1;i<n;i++)
        {
            int j = i-1;
            int temp = arr[i];
            while(j>=0 && arr[j]>temp)
            {
            //  shift element 1 place right
                arr[j+1] = arr[j];
                j--;
            }
            arr[j+1]  = temp;
        }
    }

    public static void main(String[] args) {
        int arr[] = {4,3,2,1};
        insertionSort(arr, arr.length);

        for(int k=0;k<arr.length;k++)
        {
            System.out.print(arr[k]+" ");
        }
    }
}

File: Week2/Python/SortingProblem/InsertionSort.py
"""
    The task is to complete the insert() function which is used to implement Insertion Sort.

    https://www.geeksforgeeks.org/problems/insertion-sort/1
"""

# Function to sort the list using insertion sort algorithm.    
def insertionSort(alist, n):
    """ T(c) -> O(n^2), S(c) -> O(1) """

    # unsorted list iteration
    for i in range(1, n):
        temp = alist[i]
        j = i-1
        
        insert(alist, j, temp)


def insert(alist, index, var):
    j = index
    temp = var

    # sorted list iteration, and finding elements correct place
    while j >= 0 and temp < alist[j]:
        alist[j+1] = alist[j]
        j -= 1

    # insert element at correct place
    alist[j+1] = temp



# Function to sort the list using recursive insertion sort algorithm.    
def insertionSortRecursive(alist, n):
    """ T(c) -> O(n^2), S(c) -> O(n) """
    # base case
    if n == 1:
        return
    
    # using recursion
    insertionSortRecursive(alist, n-1)
    
    i = n-1
    temp = alist[i]
    j = i-1
        
    insert(alist, j, temp)


    
arr = [5, 4, 3, 2, 1]
print(arr)
insertionSort(arr, len(arr))
print(arr)
print("-----------------------")
arr = [4, 1, 3, 9, 7]
print(arr)
insertionSort(arr, len(arr))
print(arr)